### ISSUE TITLE
improve configuration of toolchains (i.e. cross-compilation / custom toolchains)

### ISSUE THREAD

**Issue by ikcalB** [MaskRay/ccls#250]
May be similar to #138, trying to phrase it more generic:

## TL;DR
I tried fiddling aroud with `.ccls` and couldn't get ccls to accept both of:
* non-std include directory (cross compiling)
* `compile_commands.json`

I could solve either one individually, but not both.
**current workaround:** manually specifying some definitions as well as the non-std C++ library's include directory

## setup
using:
* `avr-gcc` (though this is the same for any custom toolchain - most commonly used for cross compiling)
* CMake (compile_commands.json are easily generated, as a bonus)

For embedded targets the device type is often selected by a commandline argument. This does als trigger an "automatic" definition of some sorts. (i.e. --mmcu=atmega328p -> `#define __ATmega328P__ 1`)
I once knew some more general cases, where compiler(s) might add definitions not explicitly specifed (maybe this is deprecated by now)

## possible solution
Let `.ccls` accept both `clang` and `%compile_commands.json`.

Nevertheless it sounds still useful to me, to let ccls determine flags / include dirs from the compiler.

I do understand the culprits associated with this (not to mention the work), at least gcc seems to play nice with clang - and is one of the most used compilers, especially considering embedded.

Let me know what you think, or how I can help.
Best regards from Austria

**Comment by MaskRay** [comment]
I don't know AVR but I try giving some information from my understanding of clangDriver.

```zsh
% clang -target avr -mmcu=atmega328p -E -dM -xc /dev/null -v
...
#define AVR 1
#define __AVR 1
#define __AVR_ATmega328P__ 1
#define __AVR__ 1
...
```

Compared with your avr-gcc command line, the clang command line needs `-target avr`.

You may have a working `compile_commands.json`, so you just need to create a
`.ccls` (https://github.com/MaskRay/ccls/wiki/Project-Setup#ccls-file)

```
%compile_commands.json
-target
avr
```

> By default .ccls compiler flags are applied only to files not listed in compile_commands.json. If this directive appears first in .ccls then after compile_commands.json is parsed, the rest of the .ccls arguments will be appended to the compiler flags for files found in compile_commands.json.

**Comment by ikcalB** [comment]
Thanks for your fast and thorough reply!

Specifying `-target avr` works (did not know AVR platform support was merged.)

Using a wrapper or the LSP client/IDE configuration, defeats the purpose of the per project configuration. In this case, clang support is good enough to make this work.

If querying the toolchain is really not an option, the `.ccls` file should at least provide the possibility to specify `clang` switches as well append to `%compile_commands.json` (if it does not already, which I'm not sure I understand correctly)

Another use case: Compiling vor ARM using clang - but not using the provided standard library, but instead a chip-vendor provided (optimized) one.

Best regards, Florian

**Comment by MaskRay** [comment]
> If querying the toolchain is really not an option, the `.ccls` file should at least provide the possibility to specify `clang` switches as well append to `%compile_commands.json` (if it does not already, which I'm not sure I understand correctly)

See the documentation.

* If the first line is `%compile_commands.json`, it appends flags to `compile_commands.json` entries.
* If not, it specifies complete flags for entries not covered by `compile_commands.json`
  + Variant: if `compile_commands.json` does not exist, the flags apply to every file.

Consider the following two `.ccls`:
```
clang
-target
avr
-mccu=atmega328p
-DOTHER_FLAGS
```
and
```
%compile_commands.json
-target
avr
-mccu=atmega328p
```

> not using the provided standard library

`-nostdinc`

**Comment by ikcalB** [comment]
While testing I've had the following 2 minimal examples:

`.ccls` number 1 
```
clang
-nostdinc++
-isystem/usr/lib/avr/include
```
what ofc excluded my `compile_commands.json` - yet the header `<avr/io.h>` was found. (project specific headers, as pulled in by CMakeLists.txt -> compile_commands.json were missing)

`.ccls` number 2
```
%compile_commands.json
-nostdinc++
-isystem/usr/lib/avr/include
```
seemed to ignore the `-nostinc++` (tried also with `-nostdinc`, `-I...` and every combination thereof), and as a result found project specific headers, yet `<avr/io.h>` was unresolved

**Comment by Islam0mar** [comment]
@MaskRay Is the following OK?

```
%compile_commands.json
-Wno-unused-command-line-argument
```

I am using -mcpu=cortex-m3 cross compiler.

**Comment by alifarazz** [comment]
https://github.com/MaskRay/ccls/issues/250#issuecomment-462336285
It's `-`*`mmcu`*`=atmega328p`
Else, everything works great!
